GCP VM Instance Stop Experiment

neelanjan00 opened the pull request:
GCP VM Instance Stop experiment stops the specified GCP VM instances and GKE nodes for a specified duration of time and later restarts them. The experiment accepts a GCP Service Account's credentials defined as a Kubernetes Secret, with sufficient permissions, for experimenting.

- Defined the experiment business logic and charts 
- Defined and performed integration testing for the Golang functions and packages
- Conducted alpha testing using Minikube for VM Instance target(s) and GKE node(s) targets

ispeakc0de commented on pkg/gcp/vm-instance-stop/environment/environment.go:
```diff
@@ -0,0 +1,52 @@
+	experimentDetails.Delay, _ = strconv.Atoi(common.Getenv("STATUS_CHECK_DELAY", "2"))
+	experimentDetails.Timeout, _ = strconv.Atoi(common.Getenv("STATUS_CHECK_TIMEOUT", "180"))
+	experimentDetails.TargetContainer = common.Getenv("TARGET_CONTAINER", "")
+	experimentDetails.VMInstanceName = common.Getenv("VM_INSTANCE_NAME", "")
```
Can we keep them VM_INSTANCE_NAMES and INSTANCE_ZONES? as they can contains multiple values

neelanjan00 replied:
Done

ispeakc0de commented on pkg/gcp/vm-instance-stop/types/types.go:
```diff
@@ -0,0 +1,33 @@
+	InstanceZone       string
+	ManagedNodegroup   string
+	Sequence           string
+	ActiveNodes        int
```
We are not using this anywhere. Either you can use it in place of [this](https://github.com/neelanjan00/litmus-go/blob/gcp/experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go#L22) or we can remove this

neelanjan00 replied:
Opted to remove ActiveNodes from types.go

neelanjan00 commented on chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+	// get the instance name or list of instance names
+	instanceNamesList := strings.Split(experimentsDetails.VMInstanceName, ",")
+	if len(instanceNamesList) == 0 {
+		return errors.Errorf("no instance name found to terminate")
```
Terminated has been replaced by stop in other places too.

neelanjan00 commented on chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+	return nil
+}
+
+//injectChaosInSerialMode causes VM instance termination in serial mode i.e. one after the otherwill cause VM instance termination in serial mode i.e. one after the other
```
This is now fixed.

neelanjan00 commented on chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+				events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosEngine")
+			}
+
+			//PowerOff the instance
```
PowerOff has been replaced by stop.

neelanjan00 commented on contribute/developer-guide/attributes.yaml:
```diff
@@ -0,0 +1,51 @@
+---
+name: "vm-instance-stop"
```
attributes.yaml has been deleted.

neelanjan00 commented on experiments/gcp/vm-instance-stop/charts/engine.yaml:
```diff
@@ -0,0 +1,39 @@
+apiVersion: litmuschaos.io/v1alpha1
+kind: ChaosEngine
```
done

neelanjan00 commented on experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go:
```diff
@@ -0,0 +1,248 @@
+	"github.com/sirupsen/logrus"
+)
+
+func VMInstanceStop(clients clients.ClientSets) {
```
Added the comment for the VMInstanceStop function

neelanjan00 commented on experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go:
```diff
@@ -0,0 +1,248 @@
+	})
+
+	// Calling AbortWatcher go routine, it will continuously watch for the abort signal and generate the required events and result
+	go common.AbortWatcherWithoutExit(experimentsDetails.ExperimentName, clients, &resultDetails, &chaosDetails, &eventsDetails)
```
Fixed

neelanjan00 commented on experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	// generating the event in chaosresult to marked the verdict as awaited
+	msg := "experiment: " + experimentsDetails.ExperimentName + ", Result: Awaited"
```
Fixed

neelanjan00 commented on experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go:
```diff
@@ -0,0 +1,248 @@
+			result.RecordAfterFailure(&chaosDetails, &resultDetails, failStep, clients, &eventsDetails)
+			return
+		}
+		log.Info("[Status]: VM instance is in running state (post chaos)")
```
The log has been placed outside the loop

neelanjan00 commented on experiments/gcp/vm-instance-stop/test/test.yml:
```diff
@@ -0,0 +1,66 @@
+            value: ''
+
+          # provide the chaos namespace
+          - name: CHAOS_NAMESPACE
```
Done

neelanjan00 commented:
ManagedNodegroup has been replaced with AutoScalingGroup

neelanjan00 commented:
Experiment logs for injecting chaos in two instances, test-instance and test-instance-1:

```
INFO[2021-06-29T20:39:51+05:30] Experiment Name: vm-instance-stop            
INFO[2021-06-29T20:39:51+05:30] [PreReq]: Procured the ENV for the  experiment 
INFO[2021-06-29T20:39:51+05:30] [PreReq]: Updating the chaos result of  experiment (SOT) 
INFO[2021-06-29T20:39:51+05:30] The instance information is as follows        Chaos Duration=30 Chaos Namespace=litmus Ramp Time=0 Instance Name="test-instance,test-instance-1" Sequence=parallel
INFO[2021-06-29T20:39:51+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2021-06-29T20:39:51+05:30] [Status]: No appLabels provided, skipping the application status checks 
INFO[2021-06-29T20:39:51+05:30] [Info]: The instances under chaos(IUC) are: [test-instance test-instance-1] 
INFO[2021-06-29T20:39:53+05:30] [Status]: VM instance is in running state    
INFO[2021-06-29T20:39:53+05:30] [Info]: Target instanceID list, [test-instance test-instance-1] 
INFO[2021-06-29T20:39:53+05:30] [Chaos]: Stopping the desired VM instance    
INFO[2021-06-29T20:39:54+05:30] Stopping VM instance:                         InstanceName=test-instance InstanceZone=us-central1-a
INFO[2021-06-29T20:39:54+05:30] [Chaos]: Stopping the desired VM instance    
INFO[2021-06-29T20:39:55+05:30] Stopping VM instance:                         InstanceName=test-instance-1 InstanceZone=us-central1-b
INFO[2021-06-29T20:39:55+05:30] [Wait]: Wait for VM instance 'test-instance' to get in stopped state 
INFO[2021-06-29T20:39:55+05:30] [Status]: Checking VM instance status        
INFO[2021-06-29T20:39:56+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:39:59+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:40:02+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:40:05+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:40:08+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:40:10+05:30] The instance state is STOPPING               
INFO[2021-06-29T20:40:13+05:30] The instance state is TERMINATED             
INFO[2021-06-29T20:40:15+05:30] [Wait]: Wait for VM instance 'test-instance-1' to get in stopped state 
INFO[2021-06-29T20:40:15+05:30] [Status]: Checking VM instance status        
INFO[2021-06-29T20:40:16+05:30] The instance state is TERMINATED             
INFO[2021-06-29T20:40:18+05:30] [Wait]: Waiting for chaos interval of 10s    
INFO[2021-06-29T20:40:28+05:30] [Chaos]: Starting back the VM instance       
INFO[2021-06-29T20:40:30+05:30] Starting VM instance:                         InstanceZone=us-central1-a InstanceName=test-instance
INFO[2021-06-29T20:40:30+05:30] [Chaos]: Starting back the VM instance       
INFO[2021-06-29T20:40:32+05:30] Starting VM instance:                         InstanceName=test-instance-1 InstanceZone=us-central1-b
INFO[2021-06-29T20:40:32+05:30] [Wait]: Wait for VM instance 'test-instance' to get in running state 
INFO[2021-06-29T20:40:32+05:30] [Status]: Checking VM instance status        
INFO[2021-06-29T20:40:32+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:35+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:38+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:41+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:44+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:47+05:30] The instance state is STAGING                
INFO[2021-06-29T20:40:50+05:30] The instance state is RUNNING                
INFO[2021-06-29T20:40:52+05:30] [Wait]: Wait for VM instance 'test-instance-1' to get in running state 
INFO[2021-06-29T20:40:52+05:30] [Status]: Checking VM instance status        
INFO[2021-06-29T20:40:52+05:30] The instance state is RUNNING                
INFO[2021-06-29T20:40:54+05:30] [Confirmation]:  chaos has been injected successfully 
INFO[2021-06-29T20:40:54+05:30] [Info]: The instances under chaos(IUC) are: [test-instance test-instance-1] 
INFO[2021-06-29T20:40:56+05:30] [Status]: VM instance is in running state (post chaos) 
INFO[2021-06-29T20:40:56+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2021-06-29T20:40:56+05:30] [Status]: No appLabels provided, skipping the application status checks 
INFO[2021-06-29T20:40:56+05:30] [The End]: Updating the chaos result of  experiment (EOT) 
```

neelanjan00 commented:
ChaosResult logs:

```
Name:         vm-instance-stop
Namespace:    litmus
Labels:       chaosUID=
              name=vm-instance-stop
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosResult
Metadata:
  Creation Timestamp:  2021-06-18T15:13:39Z
  Generation:          12
  Managed Fields:
    API Version:  litmuschaos.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .:
          f:chaosUID:
          f:name:
      f:spec:
        .:
        f:experiment:
      f:status:
        .:
        f:experimentStatus:
        f:history:
    Manager:         experiment
    Operation:       Update
    Time:            2021-06-18T15:13:39Z
  Resource Version:  83946
  UID:               b8d610b0-c6b4-4ea2-9715-8cbb9c8a6485
Spec:
  Experiment:  vm-instance-stop
Status:
  Experiment Status:
    Fail Step:                 N/A
    Phase:                     Completed
    Probe Success Percentage:  100
    Verdict:                   Pass
  History:
    Failed Runs:   4
    Passed Runs:   2
    Stopped Runs:  0
Events:            <none>
```

neelanjan00 commented:
ChaosResult description:

```
Name:         vm-instance-stop
Namespace:    litmus
Labels:       chaosUID=
              name=vm-instance-stop
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosResult
Metadata:
  Creation Timestamp:  2021-06-18T15:13:39Z
  Generation:          12
  Managed Fields:
    API Version:  litmuschaos.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .:
          f:chaosUID:
          f:name:
      f:spec:
        .:
        f:experiment:
      f:status:
        .:
        f:experimentStatus:
        f:history:
    Manager:         experiment
    Operation:       Update
    Time:            2021-06-18T15:13:39Z
  Resource Version:  83946
  UID:               b8d610b0-c6b4-4ea2-9715-8cbb9c8a6485
Spec:
  Experiment:  vm-instance-stop
Status:
  Experiment Status:
    Fail Step:                 N/A
    Phase:                     Completed
    Probe Success Percentage:  100
    Verdict:                   Pass
  History:
    Failed Runs:   4
    Passed Runs:   2
    Stopped Runs:  0
Events:            <none>
```

ispeakc0de commented on chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+					return errors.Errorf("unable to stop the vm instance, err: %v", err)
+				}
+
+				common.SetTargets(instanceNamesList[i], "reverted", "VM Instance Name", chaosDetails)
```
we can update chaosStatus to reverted after the VM is up. i.e line-226

neelanjan00 replied:
Done.

ispeakc0de commented on experiments/gcp/vm-instance-stop/rbac.yaml:
```diff
@@ -0,0 +1,49 @@
+  name: vm-instance-stop-sa
+  namespace: default
+  labels:
+    name: vm-instance-stop-sa
```
This SA name is correct as per naming convention (<experiment-name>-sa) but it can confuse that SA belongs to generic VM stop not limited to GCP only. Either we can keep sa name as `gcp-instance-stop-sa` or we can change the experiment name to gcp-instance-stop.

We can keep the same directory structure as the category name itself contains gcp keyword so the sub-directory doesn't require the gcp prefix but we can keep gcp prefix in the experiment name

neelanjan00 replied:
The experiment has been renamed to `gcp-vm-instance-stop` and thus the SA name has now become `gcp-vm-instance-stop-sa`. Directory names have been suitably updated too.

ispeakc0de commented on pkg/cloud/gcp/vm-instance-status.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+// PreChaosNodeStatusCheck fetches all the nodes in the cluster and checks their status, and fetches the total active nodes in the cluster, prior to the chaos experiment
+func PreChaosNodeStatusCheck(timeout, delay int, clients clients.ClientSets) (int, error) {
```
Can we reuse [this](https://github.com/litmuschaos/litmus-go/blob/master/pkg/cloud/aws/ec2/ec2-instance-status.go#L48) function? -- if you want you can move this function to https://github.com/litmuschaos/litmus-go/blob/master/pkg/utils/common/nodes.go and reuse it at other places

ispeakc0de replied:
Please check other functions as well, if we can reuse them

neelanjan00 replied:
Replaced PreChaosNodeStatusCheck, PostChaosActiveNodeCountCheck, and getActiveNodeCount functions to nodes.go; updated the gcp-vm-instance-stop, ec2-terminate-by-id, and ec2-terminate-by-tag.go experiments for the same.

uditgaurav commented on bin/experiment/experiment.go:
```diff
@@ -139,6 +140,8 @@ func main() {
 		podDNSSpoof.PodDNSSpoof(clients)
 	case "vm-poweroff":
 		vmpoweroff.VMPoweroff(clients)
+	case "vm-instance-stop":
```
We can rename the complete experiment to `gcp-instance-stop` from `vm-instance-stop`.

So we can rename the functions and directories accordingly:

VMInstanceStop() => GCPInstanceStop()
PrepareVMStop() => PrepareGCPInstanceStop() 
so on ...

Similarly,

`chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go` => `chaoslib/litmus/gcp-instance-stop/lib/gcp-instance-stop.go`
so on ...

neelanjan00 replied:
> Can we use `gcp` prefix either in import alias or in function name? to differentiate it w/ AWS, ASK instance-stop experiments?

Done

neelanjan00 replied:
> We can rename the complete experiment to `gcp-instance-stop` from `vm-instance-stop`.
> 
> So we can rename the functions and directories accordingly:
> 
> VMInstanceStop() => GCPInstanceStop()
> PrepareVMStop() => PrepareGCPInstanceStop()
> so on ...
> 
> Similarly,
> 
> `chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go` => `chaoslib/litmus/gcp-instance-stop/lib/gcp-instance-stop.go`
> so on ...

The experiment has been renamed to `gcp-vm-instance-stop` as `gcp-instance-stop` sounds a bit ambiguous since an instance can correspond to any type of instance, say, GCP Cloud SQL instance as well.

There's a bit of concern about altering the function identifiers; now that the name of the experiment is changed, if we add a `GCP` prefix to the identifiers it will result in lengthier names, and redundant as well since now the respective file names already have a `gcp-` prefix.

uditgaurav commented on chaoslib/litmus/vm-instance-stop/lib/vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+
+				// wait for the chaos interval
+				log.Infof("[Wait]: Waiting for chaos interval of %vs", experimentsDetails.ChaosInterval)
+				time.Sleep(time.Duration(experimentsDetails.ChaosInterval) * time.Second)
```
```suggestion
				common.WaitForDuration(experimentsDetails.ChaosInterval)
```

neelanjan00 replied:
Updated functions to wait for chaos interval with common.WaitForDuration() for both serial and parallel sequence of execution.

uditgaurav commented on experiments/gcp/vm-instance-stop/experiment/vm-instance-stop.go:
```diff
@@ -0,0 +1,246 @@
+		"Chaos Namespace": experimentsDetails.ChaosNamespace,
+		"Ramp Time":       experimentsDetails.RampTime,
+		"Instance Name":   experimentsDetails.VMInstanceName,
+		"Sequence":        experimentsDetails.Sequence,
```
in place of sequence we can print check zone here

neelanjan00 replied:
Pardon me; I didn't quite get you, what do you mean by "print check"?

neelanjan00 replied:
Fixed

ispeakc0de commented on chaoslib/litmus/gcp-vm-instance-stop/lib/gcp-vm-instance-stop.go:
```diff
@@ -0,0 +1,266 @@
+				//Wait for VM instance to completely stop
+				log.Infof("[Wait]: Wait for VM instance '%v' to get in stopped state", instanceNamesList[i])
+				if err := gcplib.WaitForVMInstanceDown(experimentsDetails.Timeout, experimentsDetails.Delay, experimentsDetails.AutoScalingGroup, instanceNamesList[i], experimentsDetails.GCPProjectID, instanceZonesList[i]); err != nil {
+					return errors.Errorf("unable to stop the vm instance, err: %v", err)
```
vm instance is not down

neelanjan00 replied:
Done

neelanjan00 commented:
Updated experiment logs:

```
W0712 08:47:21.767394       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2021-07-12T08:47:21Z" level=info msg="Experiment Name: gcp-vm-instance-stop"
time="2021-07-12T08:47:21Z" level=info msg="[PreReq]: Procured the ENV for the gcp-vm-instance-stop experiment"
time="2021-07-12T08:47:21Z" level=info msg="[PreReq]: Updating the chaos result of gcp-vm-instance-stop experiment (SOT)"
time="2021-07-12T08:47:21Z" level=info msg="The instance information is as follows" Ramp Time=0 Instance Names="test-instance,test-instance-1" Zones="us-central1-a,us-central1-b" Chaos Duration=30 Chaos Namespace=default
time="2021-07-12T08:47:21Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-07-12T08:47:21Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-07-12T08:47:21Z" level=info msg="[Info]: The instances under chaos(IUC) are: [test-instance test-instance-1]"
time="2021-07-12T08:47:25Z" level=info msg="[Status]: VM instance is in running state"
time="2021-07-12T08:47:25Z" level=info msg="[Info]: Target instanceID list, [test-instance test-instance-1]"
time="2021-07-12T08:47:25Z" level=info msg="[Chaos]: Stopping the desired VM instance"
time="2021-07-12T08:47:26Z" level=info msg="Stopping VM instance:" InstanceZone=us-central1-a InstanceName=test-instance
time="2021-07-12T08:47:26Z" level=info msg="[Chaos]: Stopping the desired VM instance"
time="2021-07-12T08:47:28Z" level=info msg="Stopping VM instance:" InstanceName=test-instance-1 InstanceZone=us-central1-b
time="2021-07-12T08:47:28Z" level=info msg="[Wait]: Wait for VM instance 'test-instance' to get in stopped state"
time="2021-07-12T08:47:28Z" level=info msg="[Status]: Checking VM instance status"
time="2021-07-12T08:47:29Z" level=info msg="The instance state is STOPPING"
time="2021-07-12T08:47:32Z" level=info msg="The instance state is STOPPING"
time="2021-07-12T08:47:35Z" level=info msg="The instance state is STOPPING"
time="2021-07-12T08:47:38Z" level=info msg="The instance state is STOPPING"
time="2021-07-12T08:47:41Z" level=info msg="The instance state is STOPPING"
time="2021-07-12T08:47:45Z" level=info msg="The instance state is TERMINATED"
time="2021-07-12T08:47:47Z" level=info msg="[Wait]: Wait for VM instance 'test-instance-1' to get in stopped state"
time="2021-07-12T08:47:47Z" level=info msg="[Status]: Checking VM instance status"
time="2021-07-12T08:47:48Z" level=info msg="The instance state is TERMINATED"
time="2021-07-12T08:47:50Z" level=info msg="[Wait]: Waiting for chaos interval of 30s"
time="2021-07-12T08:48:20Z" level=info msg="[Chaos]: Starting back the VM instance"
time="2021-07-12T08:48:22Z" level=info msg="Starting VM instance:" InstanceName=test-instance InstanceZone=us-central1-a
time="2021-07-12T08:48:22Z" level=info msg="[Chaos]: Starting back the VM instance"
time="2021-07-12T08:48:23Z" level=info msg="Starting VM instance:" InstanceZone=us-central1-b InstanceName=test-instance-1
time="2021-07-12T08:48:23Z" level=info msg="[Wait]: Wait for VM instance 'test-instance' to get in running state"
time="2021-07-12T08:48:23Z" level=info msg="[Status]: Checking VM instance status"
time="2021-07-12T08:48:24Z" level=info msg="The instance state is STAGING"
time="2021-07-12T08:48:28Z" level=info msg="The instance state is STAGING"
time="2021-07-12T08:48:31Z" level=info msg="The instance state is STAGING"
time="2021-07-12T08:48:34Z" level=info msg="The instance state is RUNNING"
time="2021-07-12T08:48:36Z" level=info msg="[Wait]: Wait for VM instance 'test-instance-1' to get in running state"
time="2021-07-12T08:48:36Z" level=info msg="[Status]: Checking VM instance status"
time="2021-07-12T08:48:37Z" level=info msg="The instance state is RUNNING"
time="2021-07-12T08:48:39Z" level=info msg="[Confirmation]: gcp-vm-instance-stop chaos has been injected successfully"
time="2021-07-12T08:48:39Z" level=info msg="[Info]: The instances under chaos(IUC) are: [test-instance test-instance-1]"
time="2021-07-12T08:48:41Z" level=info msg="[Status]: VM instance is in running state (post chaos)"
time="2021-07-12T08:48:41Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-07-12T08:48:41Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-07-12T08:48:41Z" level=info msg="[The End]: Updating the chaos result of gcp-vm-instance-stop experiment (EOT)"
```

neelanjan00 commented:
Updated experiment ChaosResult:

```
Name:         gcp-vm-chaos-gcp-vm-instance-stop
Namespace:    default
Labels:       app.kubernetes.io/component=experiment-job
              app.kubernetes.io/part-of=litmus
              app.kubernetes.io/version=latest
              chaosUID=50d56c83-4ac1-4526-b30b-6b64712d2afa
              controller-uid=e2846d03-94cd-4023-80d4-cb6651dceda7
              job-name=gcp-vm-instance-stop-b4imfv
              name=gcp-vm-chaos-gcp-vm-instance-stop
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosResult
Metadata:
  Creation Timestamp:  2021-07-12T08:47:21Z
  Generation:          2
  Managed Fields:
    API Version:  litmuschaos.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .:
          f:app.kubernetes.io/component:
          f:app.kubernetes.io/part-of:
          f:app.kubernetes.io/version:
          f:chaosUID:
          f:controller-uid:
          f:job-name:
          f:name:
      f:spec:
        .:
        f:engine:
        f:experiment:
      f:status:
        .:
        f:experimentStatus:
        f:history:
    Manager:         experiments
    Operation:       Update
    Time:            2021-07-12T08:47:21Z
  Resource Version:  219492
  UID:               22f867ac-f69b-42af-8afe-626a160d5e14
Spec:
  Engine:      gcp-vm-chaos
  Experiment:  gcp-vm-instance-stop
Status:
  Experiment Status:
    Fail Step:                 N/A
    Phase:                     Completed
    Probe Success Percentage:  100
    Verdict:                   Pass
  History:
    Failed Runs:   0
    Passed Runs:   1
    Stopped Runs:  0
    Targets:
      Chaos Status:  reverted
      Kind:          VM
      Name:          test-instance
      Chaos Status:  reverted
      Kind:          VM
      Name:          test-instance-1
Events:
  Type    Reason   Age    From                               Message
  ----    ------   ----   ----                               -------
  Normal  Awaited  4m54s  gcp-vm-instance-stop-b4imfv-xjlz2  experiment: gcp-vm-instance-stop, Result: Awaited
  Normal  Pass     3m34s  gcp-vm-instance-stop-b4imfv-xjlz2  experiment: gcp-vm-instance-stop, Result: Pass
```
